### PR TITLE
Create rule S6932: Use model binding instead of reading raw request data

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes https://github.com/SonarSource/sonar-dotnet/issues/8871

Replaces #8902

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.RuleDescriptorGenerator/packages.lock.json
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.RuleDescriptorGenerator/packages.lock.json:1]: This is a left over after the project was moved. It can safely be deleted.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:1]: This file seems to follow Roslyn code conventions, which is where the file is taken from, as opposed to our own.
Given that we don't have an outcome yet, for [this Trello](https://trello.com/c/nw3yLmEY/52-copying-internal-code-from-roslyn) card, I will not review this file from a stylistic point of view in this PR, but only functionally.
However, I would ask you to mention this file in the "How we work" ticket, as a possible action to to review/refactor this file, once a decision is taken on how to deal with imported code from Roslyn.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:1]: Please do not review. We will not make any edits to the file; the code is already being tested. We want to treat it as if it was a third party dll.

**Review by antonioaversa — CHANGES_REQUESTED** [review]
I am releasing a first bunch of comments, mostly on the rule, C# extensions, facade and helpers. I still need to review the rest (common, C# trackers, VB, rule tests).

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 1]
```diff
@@ -179,7 +179,6 @@ node switch
                 DestructorDeclarationSyntax { Identifier: var identifier } => identifier,
                 EnumMemberDeclarationSyntax { Identifier: var identifier } => identifier,
                 EventDeclarationSyntax { Identifier: var identifier } => identifier,
-                IdentifierNameSyntax { Identifier: var identifier } => identifier,
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:182]: `IdentifierNameSyntax` is already covered by it's base class `SimpleNameSyntax` in line 202. 

https://github.com/SonarSource/sonar-dotnet/blob/c1d3ad06b89d6ec28750e3367f027e52ab9cc023/analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs#L202

This change is "clean as you code" and unrelated to the `SimpleBaseTypeSyntax`

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 2]
```diff
@@ -199,6 +198,7 @@ node switch
                 PointerTypeSyntax { ElementType: { } elementType } => GetIdentifier(elementType),
                 PredefinedTypeSyntax { Keyword: var keyword } => keyword,
                 QualifiedNameSyntax { Right.Identifier: var identifier } => identifier,
+                SimpleBaseTypeSyntax { Type: { } type } => GetIdentifier(type),
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:201]: As long as we do not use version of Roslyn with nullable annotations, we should always assume null as a possible value. Otherwise, we a re always in the "not sure" camp and have to consult `Syntax.xml` or any other kind of documentation. Let's just play "Better safe than sorry".

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 3]
```diff
@@ -363,23 +363,14 @@ static bool TakesExpressionTree(SymbolInfo info)
...
-                result = t;
-                return true;
-            }
-            result = null;
-            return false;
-        }
-
         // based on Type="ArgumentListSyntax" in https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Syntax/Syntax.xml
-        public static ArgumentListSyntax ArgumentList(this SyntaxNode node) =>
+        public static BaseArgumentListSyntax ArgumentList(this SyntaxNode node) =>
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:367]: The comment was wrong.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 4]
```diff
@@ -363,23 +363,14 @@ static bool TakesExpressionTree(SymbolInfo info)
...
             node switch
             {
                 ObjectCreationExpressionSyntax creation => creation.ArgumentList,
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:370]: There is no `BaseObjectCreationExpressionSyntax` in our version of Roslyn. The `ImplicitObjectCreationExpressionSyntaxWrapper` case below covers that case.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 5]
```diff
@@ -363,23 +363,14 @@ static bool TakesExpressionTree(SymbolInfo info)
...
                 InvocationExpressionSyntax invocation => invocation.ArgumentList,
+                ElementAccessExpressionSyntax x => x.ArgumentList,
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:372]: You miss-read the comment. The comment hinted to our future selves to ensure the list is exhaustive. It is not meant as a "go there and follow it strictly". It means, "If you want to make your life easier, go there and use this file to find all syntax that deals with any argument list".

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 6]
```diff
@@ -503,6 +494,44 @@ public static ConditionalAccessExpressionSyntax GetRootConditionalAccessExpressi
             return current;
         }
 
+        // Copy of
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:497]: Agreed. You may want to join https://trello.com/c/nw3yLmEY We have a meeting today.
The way I did it here, is proposed but not decided yet. In this file, there are already a lot of copied parts from Roslyn and I kept that for the time being.

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpMethodParameterLookup.cs [hunk 7]
```diff
@@ -28,15 +28,18 @@ public CSharpMethodParameterLookup(InvocationExpressionSyntax invocation, Semant
...
         : base(argumentList.Arguments, methodSymbol) { }
 
     protected override SyntaxNode Expression(ArgumentSyntax argument) =>
         argument.Expression;
 
     protected override SyntaxToken? GetNameColonArgumentIdentifier(ArgumentSyntax argument) =>
         argument.NameColon?.Name.Identifier;
+
+    protected override SyntaxToken? GetNameEqualsArgumentIdentifier(ArgumentSyntax argument) =>
+        null;
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpMethodParameterLookup.cs:44]: * There is a `NameColon` property on ArgumentSyntax in CSharp.
* There are `NameColon` and `NameEquals` in `AttributeArgumentSyntax` in CSharp.
* There is a `NameColonEquals`  property in VB used for both invocation arguments and attribute arguments.

In `f(x: x1, y = x1)` the NameEqualsSyntax is within the `Expression` of the argument.

The problem is that `a: 1` (CS) or `a :=1` (VB) map to parameters, while `a = 1` (CS) `a := 1` both map to properties in attributes. Therefore, we end up with this matrix:

|  | invocation | attribute |
| --- | --- | --- |
| named argument | CS (NameColon) VB (NameColonEquals) | CS (NameColon) VB (not supported) |
| property assignment | - | CS (NameEquals) VB (NameColonEquals) |

This is reflected by the type hierarchy:

```cs
class CSharpAttributeParameterLookup
{
    protected override SyntaxToken? GetNameColonArgumentIdentifier(AttributeArgumentSyntax argument) =>
        argument.NameColon?.Name.Identifier;

    protected override SyntaxToken? GetNameEqualsArgumentIdentifier(AttributeArgumentSyntax argument) =>
        argument.NameEquals?.Name.Identifier;
}
class CSharpMethodParameterLookup
{
    protected override SyntaxToken? GetNameColonArgumentIdentifier(ArgumentSyntax argument) =>
        argument.NameColon?.Name.Identifier;

    protected override SyntaxToken? GetNameEqualsArgumentIdentifier(ArgumentSyntax argument) =>
        null;
}

class VisualBasicAttributeParameterLookup
{
    protected override SyntaxToken? GetNameColonArgumentIdentifier(ArgumentSyntax argument) =>
        null;

    protected override SyntaxToken? GetNameEqualsArgumentIdentifier(ArgumentSyntax argument) =>
        argument is SimpleArgumentSyntax { NameColonEquals.Name.Identifier: var identifier }
            ? identifier
            : null;
}
class VisualBasicMethodParameterLookup
{
    protected override SyntaxToken? GetNameColonArgumentIdentifier(ArgumentSyntax argument) =>
        (argument as SimpleArgumentSyntax)?.NameColonEquals?.Name.Identifier;

    protected override SyntaxToken? GetNameEqualsArgumentIdentifier(ArgumentSyntax argument) =>
       null;
}
```

I think this nicely reflects what is going on here. We may could do better with naming though:
* `GetNameColonArgumentIdentifier` -> `GetNamedArgumentIdentifier`
* `GetNameEqualsArgumentIdentifier` -> `GetPropertySetterArgumentIdentifier`

What do you think?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    public UseModelBinding() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context)
+    {
+        context.RegisterCompilationStartAction(compilationStartContext =>
+        {
+            var (argumentDescriptors, propertyAccessDescriptors) = GetDescriptors(compilationStartContext.Compilation);
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:42]: I think the `ValueTuple` ship has sailed. There is no need to pay any attention to it anymore.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // If the user overrides any action filters, model binding may not be working as expected. Then we do not want to raise on expressions that originate from parameters.
+                    // See the OverridesController.Undecidable test cases for details.
+                    var hasOverrides = false;
+                    var controllerCandidates = new ConcurrentStack<ReportCandidate>(); // In SymbolEnd, we filter the candidates based on the overriding we learn on the go.
+                    if (symbolStartContext.Symbol is INamedTypeSymbol namedType && namedType.IsControllerType())
+                    {
+                        symbolStartContext.RegisterCodeBlockStartAction<SyntaxKind>(codeBlockStart =>
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:53]: Without suffix works well for "Start" actions, but for other context like "NodeContext" you need the suffix. I removed the suffix from Start/End actions and kept it otherwise.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                        }
+                    });
+                }, SymbolKind.NamedType);
+            }
+        });
+    }
+
+    private bool RegisterCodeBlockActions(SonarCodeBlockStartAnalysisContext<SyntaxKind> codeBlockStart,
+        ArgumentDescriptor[] argumentDescriptors, MemberDescriptor[] propertyAccessDescriptors,
+        ConcurrentStack<ReportCandidate> controllerCandidates)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:70]: In theory, the code block registrations are executed in arbitrary order and therefore there is no order in the collection.
Therefore, I do not see how one is better than the other, but we can change them if you like.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            // We do not want to raise in ActionFilter overrides. The SymbolEndAction needs to be made aware, that there are
+            // ActionFilter overrides, so it can filter out some candidates.
+            return true;
+        }
+        // Within a single code block, access via constant and variable keys could be mixed
+        // We only want to raise, if all access were done via constants
+        var allConstantAccess = true;
+        var codeBlockCandidates = new ConcurrentStack<ReportCandidate>();
+        if (argumentDescriptors.Any())
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:82]: This is for extensibility: In the future we will have:
* No descriptors (no references to either versions)
* Descriptors for MVC 5.x
* Descriptors for Core MVC
* Maybe descriptors for MVC x.x
This solution is more flexible.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            codeBlockStart.RegisterNodeAction(nodeContext =>
+            {
+                var argument = (ArgumentSyntax)nodeContext.Node;
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:86]: It is three lines of code more and does the same thing (We do not care for the additional allocation: We use SymbolStart to limit ourselves to controllers already). It may help with readability. If you prefer it strongly I will change it but for the moment I keep it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+                var context = new ArgumentContext(argument, nodeContext.SemanticModel);
+                if (allConstantAccess && Array.Exists(argumentDescriptors, x => Language.Tracker.Argument.MatchArgument(x)(context)))
+                {
+                    allConstantAccess &= nodeContext.SemanticModel.GetConstantValue(argument.Expression) is { HasValue: true, Value: string };
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:90]: I'm very sceptical about how `FindConstantValue` is operating. I'm pretty sure it is much to naive in it's approach. I doubt that it properly handles mutation of capture:
```cs
var x = "id";
MutateTheCapture();
_ = Request.Form[x];

void MutateTheCapture() => x = "hello";
```
See `ConstantValueFinder.FindAssignedConstant` and `FindLinearPrecedingAssignmentExpression` and `CSharpAssignmentFinder.IdentifierMutation`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+                    codeBlockCandidates.Push(new(UseModelBindingMessage, GetPrimaryLocation(argument), OriginatesFromParameter(nodeContext.SemanticModel, argument)));
+                }
+            }, SyntaxKind.Argument);
+        }
+        if (propertyAccessDescriptors.Any())
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:95]: Same as above: This is the extension point for framework support in the future.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            codeBlockStart.RegisterNodeAction(nodeContext =>
+            {
+                var memberAccess = (MemberAccessExpressionSyntax)nodeContext.Node;
+                var context = new PropertyAccessContext(memberAccess, nodeContext.SemanticModel, memberAccess.Name.Identifier.ValueText);
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:100]: I think the Trackers needs some rework anyway. Look at the `PropertyAccessTrackerTest` class. That isn't good test coverage to begin with.
I added `_ = Request.Form.File\u0073;` test case

File: analyzers/src/SonarAnalyzer.RuleDescriptorGenerator/packages.lock.json [hunk 16]
```diff
@@ -1,938 +0,0 @@
-{
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.RuleDescriptorGenerator/packages.lock.json:1]: Why was this deleted?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            if (allConstantAccess)
+            {
+                controllerCandidates.PushRange([.. codeBlockCandidates]);
+            }
+        });
+        return false;
+    }
+
+    private static (ArgumentDescriptor[] ArgumentDescriptors, MemberDescriptor[] PropertyAccessDescriptors) GetDescriptors(Compilation compilation)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:117]: We likely use it and we can not prevent the use anymore. Tuple is in netstandard we target.

**Comment by costin-zaharia-sonarsource** [comment]
Hi @martin-strecker-sonarsource, this is an impressive PR. With 4788 lines changed, it is very hard to review properly. 

I can see multiple topics handled in this PR:
- Introduction of new extensions: (analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs, analyzers/src/SonarAnalyzer.Common/Helpers/SyntaxNodeExtensions.Roslyn.cs, analyzers/src/SonarAnalyzer.VisualBasic/Extensions/ExpressionSyntaxExtensions.Roslyn.cs)
- ArgumentTracker and ArgumentDescriptor
- Refactoring for the SyntaxTracker -> analyzers/src/SonarAnalyzer.Common/Trackers/SyntaxTrackerBase.cs
- Introduction of CombinatorialDataAttribute
- Test cases
- Rule implementation

I suggest splitting this into digestible pieces :) 

Regarding the implementation. I just started to review it, and I'm still wrapping my head around it, but can't we start with a simple implementation that uses a syntax walker? at least for the code blocks?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:117]: No, we're not. We had issues in the past due to it. This was the main reason we had in the past peachee projects with msbuild 14.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+        var propertyAccessDescriptors = new List<MemberDescriptor>();
+        if (compilation.GetTypeByMetadataName(KnownType.Microsoft_AspNetCore_Mvc_ControllerAttribute) is { })
+        {
+            AddAspNetCoreDescriptors(argumentDescriptors, propertyAccessDescriptors);
+        }
+        // TODO: Add descriptors for Asp.Net MVC 4.x
+        return ([.. argumentDescriptors], [.. propertyAccessDescriptors]);
+    }
+
+    private static void AddAspNetCoreDescriptors(List<ArgumentDescriptor> argumentDescriptors, List<MemberDescriptor> propertyAccessDescriptors)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:129]: From my point of view, the ArgumentDescriptor should be moved to a "KnownArguments" class and be static properties there, pretty much like "KnownMethods" and "KnownTypes". Ideally the "Known..." approach and the "Tracker..." approach are combined into a single one because they are two mechanisms that do the same.
We need a dedicated effort to achieve that.

I went for the list-based populating approach to be more flexible. e.g. if https://github.com/dotnet/aspnetcore/issues/7538 is implemented, we could at Cookies to the collection depending on a version check.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+    {
+        argumentDescriptors.AddRange([
+            ArgumentDescriptor.ElementAccess(// Request.Form["id"]
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:132]: I agree, that this is very verbose. But I do not like all the magic words in the simplified version:
* What is "Form"?
* What is `_ => true` doing?
* What is `0` referring to?

What I would like to do is create a fluent API for ArgumentDescriptor and friends instead:
`ArgumentDescriptor.ElementAccess.OfIndexerType(KnownType.Microsoft_AspNetCore_Http_IFormCollection).ForPropertyWithName("Form").WithParameterConstraint(_ => true).WithArgumentPosition(0);`

Until then, named arguments are the best we have.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+    private static (ArgumentDescriptor[] ArgumentDescriptors, MemberDescriptor[] PropertyAccessDescriptors) GetDescriptors(Compilation compilation)
+    {
+        var argumentDescriptors = new List<ArgumentDescriptor>();
+        var propertyAccessDescriptors = new List<MemberDescriptor>();
+        if (compilation.GetTypeByMetadataName(KnownType.Microsoft_AspNetCore_Mvc_ControllerAttribute) is { })
+        {
+            AddAspNetCoreDescriptors(argumentDescriptors, propertyAccessDescriptors);
+        }
+        // TODO: Add descriptors for Asp.Net MVC 4.x
+        return ([.. argumentDescriptors], [.. propertyAccessDescriptors]);
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs:126]: Why not return the lists directly?

File: analyzers/src/SonarAnalyzer.Common/Facade/SyntaxFacade.cs [hunk 21]
```diff
@@ -44,6 +44,7 @@ public abstract class SyntaxFacade<TSyntaxKind>
     public abstract bool IsAnyKind(SyntaxNode node, params TSyntaxKind[] syntaxKinds);
     public abstract bool IsAnyKind(SyntaxTrivia trivia, params TSyntaxKind[] syntaxKinds);
     public abstract bool IsInExpressionTree(SemanticModel model, SyntaxNode node);
+    public abstract bool IsWrittenTo(SyntaxNode expression, SemanticModel semanticModel, CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Facade/SyntaxFacade.cs:47]: The method originates from Roslyn and I kept parameter order also in the forwarders. Do you want me to change it, or to keep it in line with what the method is forwarded to? (We shall not change it for the methods that are copied from Roslyn).

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs [hunk 22]
```diff
@@ -0,0 +1,288 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:3]: Will be discussed in https://trello.com/c/nw3yLmEY

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+ */
+
+using System.CodeDom.Compiler;
+
+namespace SonarAnalyzer.Extensions;
+
+[GeneratedCode("Copied From Roslyn", "575bc42589145ba18b4f1cc2267d02695f861d8f")]
+public partial class ExpressionSyntaxExtensions
+{
+    // Copied from
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:28]: This is discussed in https://trello.com/c/nw3yLmEY

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs [hunk 24]
```diff
@@ -0,0 +1,288 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.CodeDom.Compiler;
+
+namespace SonarAnalyzer.Extensions;
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:23]: This will be discussed in https://trello.com/c/nw3yLmEY

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UseModelBinding_AspNetCore.cs [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+        //  ^^^^^^^^^^^^^
+        _ = Request.Query.TryGetValue("id", out _);       // Noncompliant {{Use model binding instead of accessing the raw request data}}
+        //  ^^^^^^^^^^^^^^^^^^^^^^^^^
+        _ = Request.RouteValues["id"];                    // Noncompliant {{Use model binding instead of accessing the raw request data}}
+        //  ^^^^^^^^^^^^^^^^^^^
+        _ = Request.RouteValues.TryGetValue("id", out _); // Noncompliant {{Use model binding instead of accessing the raw request data}}
+        //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        _ = Request.Form.Files;                           // Noncompliant {{Use IFormFile or IFormFileCollection binding instead}}
+        //  ^^^^^^^^^^^^^^^^^^
+        _ = Request.Form.Files["file"];                   // Noncompliant {{Use IFormFile or IFormFileCollection binding instead}}
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/UseModelBinding_AspNetCore.cs:36]: I used a parameter instead. It serves the same purpose.

**Comment by martin-strecker-sonarsource** [comment]
Superseded by 
https://github.com/SonarSource/sonar-dotnet/pull/8947
https://github.com/SonarSource/sonar-dotnet/pull/8950
#8953

#8952
#8951
#8949
#8948
